File: examples/perspective_correction.py
import numpy as np
import cv2 as cv

def mouse_event_handler(event, x, y, flags, param):
    if event == cv.EVENT_LBUTTONDOWN:
        param.append((x, y))

if __name__ == '__main__':
    img_file = '../data/sunglok_card.jpg'
    card_size = (450, 250)
    offset = 10

    # Prepare the rectified points
    pts_dst = np.array([[0, 0], [card_size[0], 0], [0, card_size[1]], [card_size[0], card_size[1]]])

    # Load an image
    img = cv.imread(img_file)
    assert img is not None, 'Cannot read the given image, ' + img_file

    # Get the matched points from mouse clicks
    pts_src = []
    wnd_name = 'Perspective Correction: Point Selection'
    cv.namedWindow(wnd_name)
    cv.setMouseCallback(wnd_name, mouse_event_handler, pts_src)
    while len(pts_src) < 4:
        img_display = img.copy()
        cv.rectangle(img_display, (offset, offset), (offset + card_size[0], offset + card_size[1]), (0, 0, 255), 2)
        idx = min(len(pts_src), len(pts_dst))
        cv.circle(img_display, offset + pts_dst[idx], 5, (0, 255, 0), -1)
        cv.imshow(wnd_name, img_display)
        key = cv.waitKey(10)
        if key == 27: # ESC
            break

    if len(pts_src) == 4:
        # Calculate planar homography and rectify perspective distortion
        H, _ = cv.findHomography(np.array(pts_src), pts_dst)
        print(f'pts_src = {pts_src}')
        print(f'pts_dst = {pts_dst}')
        print(f'H = {H}')
        img_rectify = cv.warpPerspective(img, H, card_size)

        # Show the rectified image
        cv.imshow('Perspective Distortion Correction: Rectified Image', img_rectify)
        cv.waitKey(0)

    cv.destroyAllWindows()

File: examples/perspective_correction.cpp
#include "opencv2/opencv.hpp"

void MouseEventHandler(int event, int x, int y, int flags, void* param)
{
    if (event == cv::EVENT_LBUTTONDOWN)
    {
        // Add the point to the given vector
        std::vector<cv::Point> *points_src = (std::vector<cv::Point> *)param;
        points_src->push_back(cv::Point(x, y));
        printf("A point (index: %zd) is selected at (%d, %d).\n", points_src->size() - 1, x, y);
    }
}

int main()
{
    const char* input = "../data/sunglok_card.jpg";
    cv::Size card_size(450, 250);

    // Prepare the rectified points
    std::vector<cv::Point> points_dst;
    points_dst.push_back(cv::Point(0, 0));
    points_dst.push_back(cv::Point(card_size.width, 0));
    points_dst.push_back(cv::Point(0, card_size.height));
    points_dst.push_back(cv::Point(card_size.width, card_size.height));

    // Load an image
    cv::Mat original = cv::imread(input);
    if (original.empty()) return -1;

    // Get the matched points from a user's mouse
    std::vector<cv::Point> points_src;
    cv::namedWindow("3DV Tutorial: Perspective Correction");
    cv::setMouseCallback("3DV Tutorial: Perspective Correction", MouseEventHandler, &points_src);
    while (points_src.size() < 4)
    {
        cv::Mat display = original.clone();
        cv::rectangle(display, cv::Rect(cv::Point(10, 10), card_size), cv::Vec3b(0, 0, 255), 2);
        size_t idx = cv::min(points_src.size(), points_dst.size() - 1);
        cv::circle(display, points_dst[idx] + cv::Point(10, 10), 5, cv::Vec3b(0, 255, 0), -1);
        cv::imshow("3DV Tutorial: Perspective Correction", display);
        if (cv::waitKey(1) == 27) break; // 'ESC' key: Exit
    }
    if (points_src.size() < 4) return -1;

    // Calculate planar homography and rectify perspective distortion
    cv::Mat H = cv::findHomography(points_src, points_dst);
    cv::Mat rectify;
    cv::warpPerspective(original, rectify, H, card_size);

    // Show the rectified image
    cv::imshow("3DV Tutorial: Perspective Correction", rectify);
    cv::waitKey(0);
    return 0;
}
